gh-107457: update dis documentation with changes in 3.12 (#108900)
python/cpython: Doc/library/dis.rst

      bytecode to specialize it for different runtime conditions. The
      adaptive bytecode can be shown by passing ``adaptive=True``.

   .. versionchanged:: 3.12

>       For instruction leading to a jump, a jump by 0 instruction should always

Reviewer: What is an "instruction leading to a jump"?

Author: I agree it is a bit convoluted. I am happy to use a different phrasing. Initially I went for this because of opcode like `FOR_ITER` that do not jump often.

Reviewer: I don't understand what you mean, so I can't offer how to reword it.

Author: Would the following re-wording read better ?

When the offset of a jump instruction is 0, the jump should always lead to the instruction directly following the jump instruction. Starting with 3.12, some jump instructions can be followed by a :opcode:`CACHE` instruction. When the argument of such an instruction is 0, it will jump to the first non-``CACHE`` instruction following the jump instruction. 

As a consequence, the presence of the :opcode:`CACHE` instructions is transparent for forward jumps but need to be taken into account when reasoning about backward jumps.

Reviewer: I think I see what you mean now. 

Can we just say that "the arg of a jump is the offset from the instruction that appears immediately after the JUMP instruction's CACHE entries" ?

Author: Can we agree on the following ? I would prefer to insist on the forward backward asymmetry

The argument of a jump is the offset of the target relative to the instruction that appears immediately after the jump instruction's CACHE entries. 

As a consequence, the presence of the CACHE instructions is transparent for forward jumps but need to be taken into account when reasoning about backward jumps.

Reviewer: Sure. I think "need --> needs" though.

Author: Sure

   .. versionadded:: 3.12

.. opcode:: END_SEND

   Removes the top value from the stack.
   Equivalent to ``POP_TOP``.

>    Used to clean up when a generator exits.

Reviewer 2: This is incorrect... the second item on the stack is removed, moving the top item down in its place (equivalent to `del STACK[-2]`).

Author: Thanks for catching this, I missed that going over the impl.

>    Used to clean up when a generator exit, hence the name.

Reviewer: ```suggestion
   Used to clean up when a generator exits, hence the name.
```

>    Used to clean up when a generator exit, hence the name.

Reviewer 3: ```suggestion
   Used to clean up when a generator exits.
```

.. opcode:: LOAD_SUPER_ATTR (namei)

>    This opcode implements :func:`super` both in its 0 argument form and in its
>    2 arguments form (e.g. ``super().method()``, ``super().attr`` and
>    ``super(cls, self).method()``, ``super(cls, self).attr``).

Reviewer 3: ```suggestion
   This opcode implements :func:`super`, both in its zero-argument and
   two-argument forms (e.g. ``super().method()``, ``super().attr`` and
   ``super(cls, self).method()``, ``super(cls, self).attr``).
```

   It pops three values from the stack (from top of stack down):
   - ``self``: the first argument to the current method

>    -  ``cls``: the class within which the current method was defined)

Reviewer 3: ```suggestion
   -  ``cls``: the class within which the current method was defined
```

   -  the global ``super``

   With respect to its argument, it works similarly to :opcode:`LOAD_ATTR`,
   except that ``namei`` is shifted left by 2 bits instead of 1.

   The low bit of ``namei`` signals to attempt a method load, as with

>    :opcode:`LOAD_ATTR` which results in pushing ``None`` and the loaded method.

Reviewer 3: ```suggestion
   :opcode:`LOAD_ATTR`, which results in pushing ``None`` and the loaded method.
```

.. opcode:: CALL_INTRINSIC_2

   Calls an intrinsic function with two arguments. Used to implement functionality

>    that is necessary but not performance critical::

Reviewer: ```suggestion
   that is not performance critical::
```

Reviewer: I can see you copied this from the 1-arg opcode. But I think it's redundant in both cases.

Author: I will update it

instructions:

.. versionchanged:: 3.12
   The collections now contain pseudo instructions and instrumented
   instructions as well. These are opcodes with values ``>= MIN_PSEUDO_OPCODE``

>    and ``>= MIN_INSTRUMENTED_OPCODE``.

Reviewer: MIN_PSEUDO_OPCODE > MIN_INSTRUMENTED_OPCODE. 

Do we need to mention both?

Author: Because MIN_INSTRUMENTED_OPCODE is the smallest of the two, it has to be mentioned since it is the relevant one to access only the base opcodes. Previously it was not mentioned.

Reviewer: so do we need to mention MIN_PSEUDO_OPCODE as well?

Author: Since the opcodes can be accessed in in the opcode module global, I am in favor of documenting the constants allowing to make sense of them yes.
